### ISSUE TITLE
Move Feed Actions away from init hook

### ISSUE THREAD

**Issue by growdev** [woocommerce/facebook-for-woocommerce#1632]
### Description of observed issue
The latest version of the plugin (2.0.5) hooks `SkyVerge\WooCommerce\Facebook\Products\Feed::schedule_feed_generation()` on init [here](https://github.com/facebookincubator/facebook-for-woocommerce/blob/master/includes/Products/Feed.php#L57). 

This means that this method can be fired with any frontend page request (Example: single product page).

The method uses [as_next_scheduled_action](https://github.com/facebookincubator/facebook-for-woocommerce/blob/master/includes/Products/Feed.php#L179) to check if the `wc_facebook_regenerate_feed` action is scheduled.

The problem is that this method uses a query that is very slow on sites with large number of scheduled actions. On a site with 500,000 scheduled actions rows the query for this check added 4 seconds to the transaction time for a single product page.

Here is a NewRelic trace of the next scheduled action query that dds 4,460 ms to the page load: 

<img width="999" alt="New Relic One 2020-10-13 15-56-45" src="https://user-images.githubusercontent.com/290886/95915502-d3719f00-0d6c-11eb-98e5-8bf412a68d33.png">

### Suggested change

Here are suggestions for reducing the impact this check has on frontend pages:
- Move the hook to `admin_init` so only admin 
- Add a check for `! is_admin()` and exit.
- Run this check via cron? 

We will probably be implementing something one of these on this site, but it would be nice to see this added to the plugin.

**Comment by simonporter007** [comment]
Hi @growdev!

Thanks for calling this out to us, you raise a good point! I had a quick word with our developers to talk about this. Ideally, we would be moving away from the feed file generation because of issues such as this, and the lack of flexibility that it allows us when syncing products. For instance, it's an all or nothing sync at the moment due to the update schedule in Commerce Manager.

The reason we weren't able to remove the feed sync straight away is that we had to make sure the API sync covered all cases that the feed file would be used for. The last of these was to sync the product on stock status changes which was introduced in one of the more recent versions of the plugin. As it happens, any new installs using v2.0.0+, which uses Facebooks latest connection API, will not create an update schedule for the feed file and so will not be used any way.

In the meantime, I realize this doesn't help with your request! That said, if you're running the latest version of our plugin, I would recommend removing the feed file generation altogether for the most performance. You can use a snippet such as: https://gist.github.com/410a1f2787fed34e7d4f1b9430993526#file-sv_facebook_for_woocommerce_remove_feed_file_generation_action-php which will unschedule any existing actions, and remove the init action completely.

If you do decide to use this snippet, I would also recommend checking for any existing update schedules on your [Facebook Catalog.](https://facebook.com/products). You can do this by navigating to **Catalog** > **Data Sources** > select the WooCommerce feed > **Settings** tab and then using the three-dot menu to delete any existing scheduled updates: https://cloud.skyver.ge/NQu1GnNL

Would you be looking to implement this for your large sites do you think?

Cheers!

**Comment by growdev** [comment]
Thanks @simonporter007 I'll implement the snippet and see if that helps!

**Comment by jokanane** [comment]
I tried the snippet, but experienced the following:

1) It does cancel the pending action, but once I remove the snippet, new actions get created & set to pending state to be run again every 15 minutes.

2) If I leave the snippet in place, it cancels the newly created action, but it also pollutes the action scheduler with a ton of cancelled actions.

Is that the expected behaviour? I was hoping that it cancels the action and no new actions will be created. Thanks!

WP 5.5.3, WC 4.7.0,  PHP 7.4.12, Facebook for WooCommerce 2.2.0

**Comment by hermosabeach** [comment]
Tried the snippet and getting errors as well.  Seeing the slowdown caused by this on our site.  

WordPress version 5.5.2
Current theme: Avada (version 7.1.1)
Current plugin: WooCommerce Google Product Feed (version 9.4.0)
PHP version 7.3.24-3+ubuntu20.04.1+deb.sury.org+1

Error Details
=============
An error of type E_ERROR was caused in line 458 of the file /www/wp-content/plugins/woocommerce-product-feeds/vendor/woocommerce/action-scheduler/classes/data-stores/ActionScheduler_DBStore.php. Error message: Uncaught InvalidArgumentException: Unidentified action 452226 in /www/wp-content/plugins/woocommerce-product-feeds/vendor/woocommerce/action-scheduler/classes/data-stores/ActionScheduler_DBStore.php:458
Stack trace:
#0 /www/wp-content/plugins/woocommerce-product-feeds/vendor/woocommerce/action-scheduler/functions.php(117): ActionScheduler_DBStore->cancel_action('452226')
#1 /www/wp-content/plugins/woocommerce-product-feeds/vendor/woocommerce/action-scheduler/functions.php(145): as_unschedule_action('wc_facebook_reg...', Array, 'facebook-for-wo...')
#2 /www/wp-content/plugins/code-snippets/php/snippet-ops.php(446) : eval()'d code(15): as_unschedule_all_actions('wc_facebook_reg...', Array, 'facebook-for-wo...')
#3 /wordpress/wp-includes/class-wp-hook.php(287): sv_remove_handle_generate_product_catalog_feed('')
#4 /wordpress/wp-includes/class-wp-hook.php(311): WP_Hook->apply_filters(NULL, Array)
#5 /wordpress/wp-includes/plugin.php(478): WP_Hook->do_action(Array)
#6 /wordpress/wp-settings.php(549): do_action('in

**Comment by haszari** [comment]
> Lets not close this issue after #1939 is merged as using admin_init is only a temporarily solution and there's probably a better long term solution.

It might be worth logging a new issue focusing on the wider problem, clearly defining the scope and possible impact - what do you think @danielbitzer ?
